asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py

```diff
@@ -0,0 +1,338 @@
+
+# %%
+## Keep rows where drop_row is False, i.e. remove rows based on exclusion criteria
+# sample = df[df["drop_row"] == False]
+sample = df[~df["drop_row"]]
+
+# %%
+## Conduct checks to ensure appropriate rows dropped
+
+for _col in ["cost", "tech_type", "original_epc_index"]:
+    assert sample[_col].isna().sum() == 0
+
+for _col in [
+    "cost_na",
+    "tech_type_na",
+    "original_epc_index_na",
+    "property_archetype_data_na",
+]:
+    assert sample[_col].unique() == False
+
+assert sample["tech_type"].unique() == "Air Source Heat Pump"
+
+# %% [markdown]
+# ## See distribution of removed cost data
+
+# %% [markdown]
+# Removed rows either have no EPC data or are not an ASHP.
+# We will see if the distribution of costs in removed rows is the same as that of rows which have been kept in the sample for analysis.
+
+# %%
+dropped_rows = filter_by_year.loc[~filter_by_year.index.isin(sample.index)]
+assert (len(filter_by_year) - len(sample)) == len(dropped_rows)
+
+# %%
+# In this case we'll normalise the data so they're comparable.
+
+fig, ax = plt.subplots(nrows=1, ncols=2, sharey=True, figsize=(10, 3))
+ax[0].hist(sample["cost"], density=True)
+ax[0].set_ylabel("Count")
+ax[1].hist(dropped_rows["cost"], density=True)
+for axis, title in zip(ax, ["kept", "dropped"]):
+    axis.set_title(f"Distribution of installation cost in {title} rows")
+    axis.set(xlabel="Installation cost, £")
+plt.show()
+
+# %%
+# There's not much benefit in doing this, suggest dropping.
+
+fig, ax = plt.subplots(nrows=1, ncols=2, sharey=True, figsize=(10, 3))
+ax[0].hist(sample["cost"], log=True)
+ax[0].set_ylabel("Count (log scale)")
+ax[1].hist(dropped_rows["cost"], log=True)
+for axis, title in zip(ax, ["kept", "dropped"]):
+    axis.set_title(f"Distribution of installation cost in {title} rows")
+    axis.set(xlabel="Installation cost, £")
+plt.show()
+
+# %% [markdown]
+# ## Note on comparison
+#
+# In this particular case, I'm interested in how similar/different the dropped and retained groups are in terms of cost.
+#
+# I also am not interested in the water source/ground source heat pumps at all, intuitively I know that they'll inflate the mean cost and give me a distribution with a heavier tail. Really, my key concern is: do costs for ASHP look different for installs where we don't have an EPC?
+#
+# If costs are similar, I can be relatively happy that the sample isn't massively missing something.
+#
+# I can do this in a couple of ways - 1. with statistical summaries, and 2. with visualisation.
+#
+# Descriptively the dropped rows and sample look similar - similar means, dropped rows has a higher std. deviation, which is interesting, sample again a bit higher at each quartile.
+#
+# Let's see what this actually looks like graphed out. NB I'm going to truncate the cost variable at £30,000 which is roughly the 99th percentile of the distribution, as I'm not actually that interested in the tail.
```

Comment:
Which dataset did you determine the 99th percentile from and how, for reproducibility?

Reply:
I don't think this really matters, it's just an indicative value. In reality you could choose any number that meaningfully removes the tail whilst retaining the vast majority of the data points.